docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+1. [System architecture and high-level design](#system-architecture-and-high-level-design).
+    1. [System architecture and component interaction](#system-architecture-and-component-interaction).
+1. [Detailed design](#detailed-design).
+1. [Usage scenarios and examples](#usage-scenarios-and-examples).
+1. [Other Work](#other-work).
+    1. [Related work-Simplification of retarget code](#related-work-simplification-of-retarget-code).
+    1. [Future Work-Deprecation of Stream](#future-work-deprecation-of-stream).
+1. [Roll-out Plan](#roll-out-plan).
+1. [Tools and configuration changes](#tools-and-configuration-changes).
+1. [Other information](#other-information).
+    1. [Deprecations](#deprecations).
+    1. [References](#references).
+
+
+### Revision history
+
+1.0 - Initial version - Evelyne Donnaes - 04/10/2019
+
+# Introduction
+
+### Overview and background
+
+Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functinality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class shoud be used for what purpose. Futhermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.
+
+This document collects proposals made by various contributors (see [references](#references)) and presents a simplified serial class hierarchy. The aim is to offer two public classes, one for unbuffered I/O and the other for buffered I/O, with a clear definition of which one should be used when. All other serial classes are deprecated and should be removed at a later stage. In addition, the serial classes dependencies will be optimised to only include components that provide valuable functionality.
+
+#### Serial classes in Mbed OS 5.14
+
+**Serial**
+
+`Serial` provides unbuffered I/O and is using blocking HAL calls. It only works for simple printing. 
+
+Its main limitations are:
+- It uses a mutex lock so it cannot be used from interrupts. 
+- It wastes a lot of CPU time spinning waiting for serial output and it cannot be used reliably for input because it needs interrupt-speed response on read calls but its high-level API cannot be used from interrupts.
+- It lacks buffering so it cannot be used reliably for input and output without flow control.
+- It pulls in the C library stdio system because it uses `Stream`.
+
+**UARTSerial**
+
+`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 
+
+It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it loses the ability to call printf directly.
```

Comment:
Clarify - it doesn't include built-in `printf` methods, I think you mean

Reply:
Correct. Will clarify